debian/man/podofoencrypt.1
.TH "PODOFOENCRYPT" "1" "2010-12-09" "PoDoFo" "podofoencrypt"
.PP
.SH NAME
podofoencrypt \- encrypt PDF files and set PDF security settings
.PP
.SH SYNOPSIS
\fBpodofoencrypt\fR [\-\-rc4v1] [\-\-rc4v2] [\-\-aes] [\-u <userpassword>] \-o <ownerpassword> <inputfile> <outputfile>
.PP
.SH DESCRIPTION
.B podofoencrypt
is one of the command line tools from the PoDoFo library that provide several
useful operations to work with PDF files\. It can encrypt PDF files using RC4
or AES encoding and can set PDF security settings\.
.PP
.SH "OPTIONS"
\fB\-\-help\fR
.RS
Display the help text
.RE
.PP
Algorithm:
.PP
.RS
\fB\-\-rc4v1\fR
.RS
Use rc4v1 encryption
.RE
.PP
\fB\-\-rc4v2\fR
.RS
Use rc4v2 encryption (Default value)
.RE
.PP
\fB\-\-aes\fR
.RS
Use aes encryption (currently not supported)
.RE
.RE
.PP
Passwords:
.PP
.RS
\fB\-u <password>\fR
.RS
An optional user password
.RE
.PP
\fB\-o <password>\fR
.RS
The required owner password
.RE
.RE
.PP
Permissions:
.RS
.PP
\fB\-\-print\fR
.RS
.PP
Allow printing the document
.RE
.PP
\fB\-\-edit\fR
.RS
.PP
Allow modification of the document besides annotations, form fields or changing pages
.RE
.PP
\fB\-\-copy\fR
.RS
.PP
Allow extraction of text and graphics
.RE
.PP
\fB\-\-editnotes\fR
.RS
.PP
Add or modify text annotations or form fields (if ePdfPermissions_Edit is set also allow the creation of interactive form fields including signature)
.RE
.PP
\fB\-\-fillandsign\fR
.RS
.PP
Fill in existing form or signature fields
.RE
.PP
\fB\-\-accessible\fR
.RS
.PP
Extract text and graphics to support user with disabillities
.RE
.PP
\fB\-\-assemble\fR
.RS
.PP
Assemble the document: insert, create, rotate delete pages or add bookmarks
.RE
.PP
\fB\-\-highprint   \fR
.RS
.PP
Print a high resolution version of the document
.PP
.SH SEE ALSO
.BR podofobox (1),
.BR podofocolor (1),
.BR podofocountpages (1),
.BR podofocrop (1),
.BR podofogc (1),
.BR podofoimg2pdf (1),
.BR podofoimgextract (1),
.BR podofoincrementalupdates (1),
.BR podofoimpose (1),
.BR podofomerge (1),
.BR podofopages (1),
.BR podofopdfinfo (1),
.BR podofotxt2pdf (1),
.BR podofotxtextract (1),
.BR podofouncompress (1),
.BR podofoxmp (1)
.PP
.SH AUTHORS
.PP
PoDoFo is written by Dominik Seichter <domseichter@web\.de> and others\.
.PP
This manual page was written by Oleksandr Moskalenko <malex@debian\.org> for
the Debian Project (but may be used by others)\.


ChangeLog
Version 0.8
	See SVN ChangeLog

Version 0.7
    Fixed PdfPainter::ArcTo
    Fixed crash in FlateDecode and LZWDecode predictor functions
    Fixed writing of unicode PdfStrings with brackets (will be escaped
	now)
    Fixed encoding of bytes in PdfName
    Added many new unit tests
    Added methods to free object memory of objects that are not needed
	anymore (These objects will be re-read from disk if they are
	needed again)
    Fixed a crash when appending PDFs
    Added unicode support on Win32 (wchar_t* constructors and methods
	where apropriate, e.g. file handling)
    Fixed a memory leak in PdfStream
    Small optimizations in various places
    Fixed DCTDecode filter for CMYK images
    PdfReference is now immutable, which allows for various optimizations
    Fixed PdfInputDevice::Read to return now the correct number of bytes
	read
    Fixed a memory leak in PdfImmediateWriter/PdfStreamedDocument
    Fixed several minor parsing issues
    Fixed adding text to existing pages
    Added Lua5.1 plan file support to podofoimpose
    Several fixes in creating XObjects from pages
    Fixed a possible crash in PdfNamesTree
    Added support for setting colors in PdfAnnotations
    Added advanced text drawing support to PdfPainter
    Fixed parsing of Type1 fonts
    Fixed deletion of PdfAnnations (fixes a memory leak)
    

Version 0.6
    Many many changes :) See SVN log

Version 0.5
    Added support for Embedded Files (annotations & named objects)
    Added support for ExtGStates when drawing
		initially only supports basic transparency
    Fixed reading values from nametrees
    Added support for named destinations
    Fixed a memory leak in PdfDestination::GetPage
    Pages do not know their page number inside of the document
    Fixed reading PdfActions from PDF files
    Moved filter implementations to PdfFiltersPrivate.h
    Added PdfFilter::CanEncode and PdfFilter::CanDecode
    Simpliefid PoDoFos handling of XRef tables
	
Version 0.4
	PdfImage now supports creating an image stream from a "raw bitmap"
		which can also be optionally Flate compressed
	Added some new Page-related methods to PdfDocument 
		Append - append one document to another
		InsertPages - insert a range of pages from one document to another
		DeletePages - delete a range of pages
	Added new tool podofomerge for merging two PDFs together
	Added methods to get & set a document's PageMode
	Added methods to set a document's FullScreen mode
	Added methods to set all the various ViewerPreferences for a document
	Added methods to set the document's PageLayout
	Added Outline support
		modified podofoinfo to display them, if present	
	Added a PdfDestination class
	Added PdfNamesTree class for handling the global named objects
	        modified podofoinfo to display them, if present
	PdfPainter can draw bezier curves
	Added XCode project for building on Mac OS X
	        fixed up conditionals in font code to enable building on
		MacOSX - but no font loading, YET
	Added support for writing linearized PDF
	Added support for garbage collection of unused objects
	Simplified PdfObject constructor
	Improved annotation support
	Added support to encode names and various name testcases
	Fixed ascent and descent of fonts
	Improved PdfImage API
	Added support for the creation of file identifiers which makes
		PoDoFo created PDF files work in more different PDF
		workflows
	PdfImage optionally takes ownership of buffers
	Fixed a major parser bug: The parser expected all objects in
		object streams to be of type dictionary.
	
Version 0.3
    TOTAL revamp of PdfObject & PdfVariant to enable clean/consistent object handling;
    	Added new PdfDocument object - new high level object for reading & writing a PDF doc;
	TOTAL revamp of the PdfDocument, PdfWriter & PdfParser relationship
		PdfDocument is now hub for both reading and writing a document
		it holds the PdfVecObjects - the others just reference them.
	TOTAL revamp of PdfPainter
		now uses PDF coordinates - UserUnits from bottom/left
			added PdfPainterMM for mm-based coords
		supports user-specified float precision
			and writes out floats in an optimal manner
		supports "appending" mode for drawing on existing documents
	Improved handling of the /Info dict for both reading and writing PDFs;
	Added new test app - podofopdfinfo, which will be used to dump metadata, etc. from a PDF;
	Added PdfError::DebugMessage() as the official way to write out debugging info;
		updated all other debugging msgs to use this;
	Added PdfError::DebugEnabled() to enable/disable display of debug messages;
	Added tracking of file size in PdfParser;
	Minor tweak to Linearization handling - to enable getting the status from a doc;
	Added getting GetPdfVersionAsString() to PdfWriter;
	Added new info/object getting methods to PdfDocument;
		bool IsLinearized();
		size_t FileSize();
		PdfObject* GetStructTreeRoot();
		PdfObject* GetMetadata();
		PdfObject* GetOutlines();
		PdfObject* GetAcroForm();
	Updated pdfinfo & podofopdfinfo to call the new PdfDocument methods;
	Added PdfDictionary and PdfArray classes;
	Added new PdfPagesTree (inside of PdfDocument.cpp) for handling walking a /Pages tree;
	Added new GetPageCount() method to PdfDocument;
	Modifications to PdfPage to attach it to a PdfDocument & construct from a PdfObject;
	Added new Legal and A3 standard page sizes;
	Changed page coordinates to be PDF unit-based instead of 1/1000mm;
	Changed PdfRect to use PDF units and also use bottom instead of top;
	Added ability to go between PdfRect and PdfArray & also get string version of a PdfRect;
	Added support for PdfPage to return all the standard boxes (Media, Crop, etc);
	Added support for fetching inherited values from pages (eg. boxes, rotation, etc.)
	Added more methods to PdfPage;
		GetRotation();
		GetNumAnnots();
	Use Exceptions now instead of error codes;
	Removed Init from PdfOutputDevice;
	Removed Init from PdfParser;
	Added LZW Filter support;

tools/podofoencrypt/podofoencrypt.cpp
/***************************************************************************
 *   Copyright (C) 2005 by Dominik Seichter                                *
 *   domseichter@web.de                                                    *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <podofo.h>

using namespace PoDoFo;

#include <cstdlib>
#include <cstdio>

#ifdef _HAVE_CONFIG
#include <config.h>
#endif // _HAVE_CONFIG

void encrypt( const char* pszInput, const char* pszOutput, 
              const std::string & userPass, const std::string & ownerPass,
              const PdfEncrypt::EPdfEncryptAlgorithm eAlgorithm, const int nPermissions ) 
{
    PdfVecObjects objects;
    PdfParser     parser( &objects );
    
    objects.SetAutoDelete( true );
    parser.ParseFile( pszInput );
    
    PdfEncrypt::EPdfKeyLength eKeyLength;
    EPdfVersion   eVersion;
    switch( eAlgorithm ) 
    {
#ifndef OPENSSL_NO_RC4
        case PdfEncrypt::ePdfEncryptAlgorithm_RC4V1:
            eKeyLength = PdfEncrypt::ePdfKeyLength_40;
            eVersion   = ePdfVersion_1_3;
            break;
#endif // OPENSSL_NO_RC4
#ifdef PODOFO_HAVE_LIBIDN
        case PdfEncrypt::ePdfEncryptAlgorithm_AESV3:;
            eKeyLength = PdfEncrypt::ePdfKeyLength_256;
            eVersion   = ePdfVersion_1_7;
            break;
#endif // PODOFO_HAVE_LIBIDN
#ifndef OPENSSL_NO_RC4
        case PdfEncrypt::ePdfEncryptAlgorithm_RC4V2:
#endif // OPENSSL_NO_RC4
        case PdfEncrypt::ePdfEncryptAlgorithm_AESV2:
        default:
            eKeyLength = PdfEncrypt::ePdfKeyLength_128;
            eVersion   = ePdfVersion_1_5;
            break;
    }

    PdfWriter writer( &parser );
	PdfEncrypt *encrypt = PdfEncrypt::CreatePdfEncrypt( userPass, ownerPass, nPermissions,
                        eAlgorithm, eKeyLength  );
    
    writer.SetPdfVersion( eVersion );
    writer.SetEncrypted( *encrypt );
    writer.Write( pszOutput );

	delete encrypt;
}

void print_help()
{
    printf("Usage: podofoencrypt [--rc4v1] [--rc4v2] [--aesv2] [--aesv3] [-u <userpassword>]\n");
    printf("                     -o <ownerpassword> <inputfile> <outputfile>\n\n");
    printf("       This tool encrypts an existing PDF file.\n\n");
    printf("       --help        Display this help text\n");
    printf(" Algorithm:\n");
    printf("       --rc4v1       Use rc4v1 encryption\n");
    printf("       --rc4v2       Use rc4v2 encryption (Default value)\n");
    printf("       --aesv2       Use aes-128 encryption\n");
    printf("       --aesv3       Use aes-256 encryption\n");
    printf(" Passwords:\n");
    printf("       -u <password> An optional userpassword\n");
    printf("       -o <password> The required owner password\n");
    printf(" Permissions:\n");
    printf("       --print       Allow printing the document\n");
    printf("       --edit        Allow modifying the document besides annotations, form fields or chaning pages\n");
    printf("       --copy        Allow text and graphic extraction\n");
    printf("       --editnotes   Add or modify text annoations or form fields (if ePdfPermissions_Edit is set also allow to create interactive form fields including signature)\n");
    printf("       --fillandsign Fill in existing form or signature fields\n");
    printf("       --accessible  Extract text and graphics to support user with disabillities\n");
    printf("       --assemble    Assemble the document: insert, create, rotate delete pages or add bookmarks\n");
    printf("       --highprint   Print a high resolution version of the document\n");
    printf("\n\n");
}
       
int main( int argc, char* argv[] )
{
  const char*                      pszInput   = NULL;
  const char*                      pszOutput  = NULL;
  PdfEncrypt::EPdfEncryptAlgorithm eAlgorithm = PdfEncrypt::ePdfEncryptAlgorithm_AESV2;
  int                              nPerm      = 0;
  std::string                      userPass;
  std::string                      ownerPass;

  if( argc < 3 )
  {
    print_help();
    exit( -1 );
  }

  // Parse the commandline options
  for( int i=1;i<argc;i++ ) 
  {
      if( argv[i][0] == '-' ) 
      {
#ifndef OPENSSL_NO_RC4
          if( strcmp( argv[i], "--rc4v1" ) == 0 ) 
              eAlgorithm = PdfEncrypt::ePdfEncryptAlgorithm_RC4V1;
          else if( strcmp( argv[i], "--rc4v2" ) == 0 ) 
              eAlgorithm = PdfEncrypt::ePdfEncryptAlgorithm_RC4V2;
          else
#endif // OPENSSL_NO_RC4
          if( strcmp( argv[i], "--aesv2" ) == 0 ) 
              eAlgorithm = PdfEncrypt::ePdfEncryptAlgorithm_AESV2;
#ifdef PODOFO_HAVE_LIBIDN
          else if( strcmp( argv[i], "--aesv3" ) == 0 ) 
              eAlgorithm = PdfEncrypt::ePdfEncryptAlgorithm_AESV3;
#endif // PODOFO_HAVE_LIBIDN
          else if( strcmp( argv[i], "-u" ) == 0 ) 
          {
              ++i;
              if( i < argc ) 
                  userPass = argv[i];
              else
              {
                  fprintf( stderr, "ERROR: -u given on the commandline but no userpassword!\n");
                  exit( -1 );
              }
          }
          else if( strcmp( argv[i], "-o" ) == 0 ) 
          {
              ++i;
              if( i < argc ) 
                  ownerPass = argv[i];
              else
              {
                  fprintf( stderr, "ERROR: -o given on the commandline but no ownerpassword!\n");
                  exit( -1 );
              }
          }
          else if( strcmp( argv[i], "--help" ) == 0 ) 
          {
              print_help();
              exit( -1 );
          }
          else if( strcmp( argv[i], "--print" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_Print;
          else if( strcmp( argv[i], "--edit" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_Edit;
          else if( strcmp( argv[i], "--copy" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_Copy;
          else if( strcmp( argv[i], "--editnotes" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_EditNotes;
          else if( strcmp( argv[i], "--fillandsign" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_FillAndSign;
          else if( strcmp( argv[i], "--accessible" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_Accessible;
          else if( strcmp( argv[i], "--assemble" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_DocAssembly;
          else if( strcmp( argv[i], "--highprint" ) == 0 ) 
              nPerm |= PdfEncrypt::ePdfPermissions_HighPrint;
          else
          {
              fprintf( stderr, "WARNING: Do not know what to do with argument: %s\n", argv[i] );
          }
      }
      else
      {
          if( !pszInput )
          {
              pszInput = argv[i];
          }
          else if( !pszOutput )
          {
              pszOutput = argv[i];
          }
          else
          {
              fprintf( stderr, "WARNING: Do not know what to do with argument: %s\n", argv[i] );
          }

      }
  }

  // Check for errors in the commandline options
  if( !pszInput ) 
  {
      fprintf( stderr, "ERROR: No input file specified\n");
      exit( -1 );
  }

  if( !pszOutput )
  {
      fprintf( stderr, "ERROR: No output file specified\n");
      exit( -1 );
  }

  if( !ownerPass.length() )
  {
      fprintf( stderr, "ERROR: No owner password specified\n");
      exit( -1 );
  }
      

  // Do the actual encryption
  try {
      encrypt( pszInput, pszOutput, userPass, ownerPass, eAlgorithm, nPerm );
  } catch( PdfError & e ) {
      fprintf( stderr, "Error: An error %i ocurred during encrypting the pdf file.\n", e.GetError() );
      e.PrintErrorMsg();
      return e.GetError();
  }


  printf("%s was sucessfully encrypted to: %s\n", pszInput, pszOutput );
  
  return 0;
}

